mindmeld/text_preparation/text_preparation_pipeline.py

```diff
@@ -0,0 +1,530 @@
+                For Example:
+                    norm_token = {
+                        "entity": "order",
+                        "raw_entity": "order",
+                        "raw_token_index": 1,
+                        "raw_start": 1
+                    }`
+        """
+        normalized_tokens = []
+        for i, raw_token in enumerate(raw_tokens):
+            if not raw_token["text"]:
+                continue
+            normalized_texts = self._normalize_single_token(raw_token["text"])
+            if len(normalized_texts) > 0:
+                for token_text in normalized_texts:
+                    normalized_tokens.append(
+                        {
+                            "entity": token_text,
+                            "raw_entity": raw_token["text"],
+                            "raw_token_index": i,
+                            "raw_start": raw_token["start"],
+                        }
+                    )
+        return normalized_tokens
+
+    def _normalize_single_token(self, text):
+        """Normalize an individual token by processing text with all normalizers.
+
+        Args:
+            text (str): Text to normalize.
+        Returns:
+            normalized_texts (List[str]): Normalized texts. We tokenize the post-norm text and
+                split the entity if possible.
+        """
+        normalized_text = text
+        for normalizer in self.normalizers:
+            normalized_text = TextPreparationPipeline.modify_around_annotations(
+                text=normalized_text, function=normalizer.normalize
+            )
+        # We tokenize the post-norm text and split the entity if possible
+        # Ex: normalize("o'clock") -> "o clock" -> ["o", "clock"]
+        return [t["text"] for t in self.tokenize(normalized_text)]
+
+    def tokenize(self, text):
+        """
+        Args:
+            text (str): Input text.
+        Returns:
+            tokens (List[str]): List of tokens.
+        """
+        return TextPreparationPipeline.tokenize_around_mindmeld_annotations(
+            text=text, function=self.tokenizer.tokenize
+        )
+
+    def stem_word(self, word):
+        """
+        Gets the stem of a word. For example, the stem of the word 'fishing' is 'fish'.
+
+        Args:
+            words (List[str]): List of words to stem.
+
+        Returns:
+            stemmed_words (List[str]): List of stemmed words.
+        """
+        return self.stemmer.stem_word(word)
+
+    @staticmethod
+    def find_mindmeld_annotation_re_matches(text):
+        """
+        Args:
+            text (str): The string to find mindmeld annotation instances
+                (" {entity_text|entity_type} ")
+        Returns:
+            matches (List[sre.SRE_Match object]): Regex match objects.
+        """
+        return list(MINDMELD_ANNOTATION_PATTERN.finditer(text))
+
+    @staticmethod
+    def modify_around_annotations(text, function):
```

Comment:
I think `modify_around_annotations ` and `tokenize_around_mindmeld_annotations ` can be combined. The main difference is the append vs extend and offset ops. The offset ops can be composed into the function that is passed in.

Reply:
`modify_around_annotations` is also used by normalizers so it may be better to keep them separate